parsons/databases/database_connector.py

```diff
@@ -0,0 +1,108 @@
+from abc import ABC, abstractmethod
+from typing import Optional
+from parsons.etl.table import Table
+
+
+class DatabaseConnector(ABC):
+    """
+    An abstract base class that provides a uniform interface for all Parsons database connectors.
+    This class should be used in functions instead of the specific database connector classes
+    when the functions don't rely on database-specific functionality.
+
+    It ensures that any class that inherits from it implements the `table_exists`, `copy`, and
+    `query` methods, which are common operations when working with databases.
+
+    Note:
+        The Python type system (as of 3.10.6) will not stop you from breaking the type contract
+        of method signatures when implementing a subclass. It is up to the author of a database
+        connector to ensure that it satisfies this interface. Be careful to, for example, not
+        change the types of the parameters or leave out optional parameters that are specified
+        in the interface.
+
+        Any such inconsistencies can cause unexpected runtime errors that will not be caught by
+        the type checker.
+
+        It is possible to safely add additional features, such as new methods or extra **optional**
+        parameters to specified methods. In general adding new methods is safe, but adding optional
+        parameters to methods specified in the interface should be considered bad practice, because
+        it could result in unexpected behavior.
+
+    Example usage:
+
+    .. code-block:: python
+
+        def my_function(db: DatabaseConnector, data: Table):
+            # Your code here, using the db object
+
+        # Pass an instance of a class that inherits from DatabaseConnector, e.g. Redshift
+        my_function(some_db_instance, some_data)
+
+    """
```

Comment:
I think we should also specify the arguments for the init method. I glanced at a few of the database connector classes and seems the all have username, password, db, host, port. Specifying that in the ABC will also make the consistency easier for the instantiation of the generic db connector, particularly like in the toy example I put in another comment.

I think the goal of being able to point the same pipeline or script at a different warehouse without a deployment (ie without changing the code) and just adjusting env variables is easier if we keep this consistent as well

Reply:
I don't think this is a good idea. Constructors in class hierarchies are special cases. That would limit the possible definition of what could be a database connector to things that only have a username, password, db, host, and port. For example, that would rule out creating a SQLite connector, because SQLite does not have any of those things.

I'm not even sure that Python's ABC's support this.

Reply:
Python's ABCs do support this (see [this StackOverflow question](https://stackoverflow.com/questions/72643177/how-to-create-a-class-with-an-abstract-init-method)) but I agree with Jason that the benefits of standardizing the `__init__` method are outweighed by the fact doing so would exclude potential edge cases like SQLite. 

However - and this may be what Willy was suggesting in the first place - we _can_ specify a default `__init__` implementation that subclasses can inherit or choose to override, without making it an abstract method that they would have to signature-match (see [this other StackOverflow question](https://stackoverflow.com/a/71159174)).

Reply:
Ah, that's good to know. Thanks for finding that, Shauna. I'm still not sure what the purpose would be, though. If the goal is just to have a documented preferred signature for a typical connector (while not ruling out an atypical case) and to save some typing, that seems fine to me.

I think that @willyraedy's vision for the `discover_database` function might be informing his comment. In his example he posted in that thread, the discover function supported passing arguments through to the different constructors. If that's the starting point, then you would _need_ the constructor signatures to have consistent parameter names and types.

Reply:
I have a modest preference for including the default/overridable implementation in `__init__`, totally okay with being overridden though.

Reply:
@willyraedy can you weigh in here and on the other comment threads in this PR? I feel like we're getting close to the finish line.

Reply:
Yeah I think this is about how implicit / explicit the multi-database support is.

My instinct is to keep it simple and as-is until we start working some of those use cases in a more concrete way.